Test fix/post merge follow up

agibsonccc opened the pull request:
## What changes were proposed in this pull request?

Fix left over build issues from merging last series of PRs
Fixes up cuda issues (mainly left over int -> long issues with kernels)
Cleans up commented code

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

## Quick checklist

The following checklist helps ensure your PR is complete:

- [ X] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.konduit.ai/multi-project/how-to-guides/contribute/eclipse-contributors) page for details
- [X ] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [ X] Created tests for any significant new code additions.
- [X ] Relevant tests for your changes are passing.

agibsonccc committed:
Extra refactoring post merge

agibsonccc committed:
Remove extra grid executioner

agibsonccc committed:
More extra cuda changes int -> long

agibsonccc committed:
Fix cpu op executioner signatures

agibsonccc committed:
Update parallelwrapper directory layout/dependencies

agibsonccc committed:
Fix JVM crashes
Fix workspace assertions

agibsonccc committed:
Remove backward from cuda build

agibsonccc committed:
Fix stack traces in cuda
Remove c++ 17 requirement for backtrace
Fix static_cast compilation error for nvcc
Add new throw exception macro when  back trace is enabled to print backtraces
when an exception is thrown. This makes debugging more accessible.
Remove shape buffer deletion which caused crashes

agibsonccc committed:
Add back dependency for lz4
Fix more cuda dimension int -> long kernel launches

agibsonccc committed:
Fix static cast on templated get
Update dilation int -> long

agibsonccc committed:
Fix diemensions uage in reduce ops
Migrate more int dimensions to longs

agibsonccc committed:
Update segment ops to use longs
Update base data buffer equals method to use c++ op

agibsonccc committed:
Update convolutions (cpu/gpu) changing indices from ints to longs

treo commented on libnd4j/include/array/NDArray.h:
```diff
@@ -1594,7 +1594,7 @@ bool NDArray::isAttached() { return this->_context->getWorkspace() != nullptr; }
-  auto v = static_cast<R>(b[index]);
+  auto v = static_cast<T>(b[index]);
   return v;
 }
```
This looks wrong. Casting to T and then still having R as return shouldn't work, unless T and R are the same everywhere.

treo commented on libnd4j/include/legacy/cuda/NativeOps.cu:
```diff
@@ -211,55 +222,86 @@ void copyBuffer(OpaqueDataBuffer *target, long n,  OpaqueDataBuffer *from, long
+collect2: error: ld returned 1 exit status
 
-SD_LIB_EXPORT int contextNumOutputs(void *contextPointer) {
+ */
```
What's with this todo?

treo commented on platform-tests/src/test/java/org/eclipse/deeplearning4j/nd4j/linalg/workspace/BasicWorkspaceTests.java:
```diff
@@ -340,7 +339,7 @@ public void testCreateDetached1(Nd4jBackend backend) {
             long reqMemory = wsI.requiredMemoryPerArray(array1);
             //multiply by 2 for the dup() calling assign
-            assertEquals(reqMemory * 2 , wsI.getPrimaryOffset());
+            assertEquals(reqMemory , wsI.getPrimaryOffset());
```
Now the comment is wrong.

agibsonccc committed:
Update convolutions (cpu/gpu) changing indices from ints to longs
Remove commented code
Update ubyte cuda buffer case
Clean up opaque buffer usage in cuda executioner

agibsonccc committed:
Update nits

treo commented on nd4j/nd4j-backends/nd4j-backend-impls/nd4j-cuda/src/main/java/org/nd4j/linalg/jcublas/buffer/BaseCudaDataBuffer.java:
```diff
@@ -671,25 +673,24 @@ public void copyDataFromSrc(Pointer pointer,long length, long srcOffset,long dst
+
+            case UBYTE: {
+                data = ArrayUtil.cutBelowZero(data);
+            }
```
Is this explicitly meant to fall through to the next case? If so, put a note here.

treo commented on nd4j/nd4j-backends/nd4j-backend-impls/nd4j-cuda/src/main/java/org/nd4j/linalg/jcublas/buffer/BaseCudaDataBuffer.java:
```diff
@@ -747,27 +763,32 @@ public void set(long[] data, long length, long srcOffset, long dstOffset) {
+
+                syncDeviceAndHost();
+
+            }
```
Is this explicitly meant to fall through to the next case? If so, put a note here.

treo commented on nd4j/nd4j-backends/nd4j-backend-impls/nd4j-cuda/src/main/java/org/nd4j/linalg/jcublas/buffer/BaseCudaDataBuffer.java:
```diff
@@ -835,25 +847,32 @@ public void set(long[] data, long length, long srcOffset, long dstOffset) {
-                }
+                data = ArrayUtil.cutBelowZero(data);
+
+            }
```
Is this explicitly meant to fall through to the next case? If so, put a note here.

treo commented on nd4j/nd4j-backends/nd4j-backend-impls/nd4j-cuda/src/main/java/org/nd4j/linalg/jcublas/buffer/BaseCudaDataBuffer.java:
```diff
@@ -907,26 +928,179 @@ public void set(float[] data, long length, long srcOffset, long dstOffset) {
             break;
+            case UBYTE: {
+                data = ArrayUtil.cutBelowZero(data);
+            }
```
Is this explicitly meant to fall through to the next case? If so, put a note here.

treo commented on nd4j/nd4j-backends/nd4j-backend-impls/nd4j-cuda/src/main/java/org/nd4j/linalg/jcublas/buffer/BaseCudaDataBuffer.java:
```diff
@@ -907,26 +928,179 @@ public void set(float[] data, long length, long srcOffset, long dstOffset) {
-                    put(e, data[e]);
-                }
+                data = ArrayUtil.cutBelowZero(data);
+            }
```
Is this explicitly meant to fall through to the next case? If so, put a note here.

treo commented on nd4j/nd4j-backends/nd4j-backend-impls/nd4j-cuda/src/main/java/org/nd4j/linalg/jcublas/buffer/BaseCudaDataBuffer.java:
```diff
@@ -949,6 +1123,91 @@ public void set(double[] data, long length, long srcOffset, long dstOffset) {
+
+            case UBYTE: {
+                data = ArrayUtil.cutBelowZero(data);
+            }
```
Is this explicitly meant to fall through to the next case? If so, put a note here.

treo reviewed (changes requested):
A few more nit picks about making it explicit that you want the behavior that you're using.

agibsonccc committed:
Fix conv2d signatures
Remove commented code

agibsonccc committed:
Handle conversion case for cuda with float16 and bfloat16 to resolve ambiguities
Resolve one more conversion type with cuda data buffers
